charmcraft/config.py

```diff
@@ -0,0 +1,175 @@
+    'object': 'dict',
+    'array': 'list',
+}
+
+
+def adapt_validation_error(error):
+    """Take a jsonschema.ValidationError and create a proper CommandError."""
+    field_path = '.'.join(error.absolute_path)
+    if error.validator == 'required':
+        msg = "Bad charmcraft.yaml content; missing fields: {}.".format(
+            ', '.join(error.validator_value))
+    elif error.validator == 'type':
+        expected_type = TYPE_TRANSLATOR.get(error.validator_value, error.validator_value)
+        msg = "Bad charmcraft.yaml content; the '{}' field must be a {}: got '{}'.".format(
+            field_path, expected_type, error.instance.__class__.__name__)
+    elif error.validator == 'enum':
+        msg = "Bad charmcraft.yaml content; the '{}' field must be one of: {}.".format(
+            field_path, ', '.join(map(repr, error.validator_value)))
+    elif error.validator == 'format':
+        msg = "Bad charmcraft.yaml content; the '{}' field {}: got {!r}.".format(
+            field_path, error.cause, error.instance)
+    else:
+        # safe fallback
+        msg = error.message
+
+    raise CommandError(msg)
+
+
+@format_checker.checks('url', raises=ValueError)
+def _check_url(value):
+    """Check that the URL has at least scheme and net location."""
+    if isinstance(value, str):
+        url = urlparse(value)
+        if url.scheme and url.netloc:
+            return True
+    raise ValueError("must be a full URL (e.g. 'https://some.server.com')")
+
+
+@attr.s(kw_only=True, frozen=True)
+class _CharmhubConfig:
+    """Configuration for all Charmhub related options."""
+
+    api_url = attr.ib(default='https://api.staging.charmhub.io')
+    storage_url = attr.ib(default='https://storage.staging.snapcraftcontent.com')
+
+    @classmethod
+    def from_dict(cls, source):
+        """Build from a raw dict."""
+        return cls(**source)
+
+
+class _BasicPrime(tuple):
+    """Hold the list of files to include, specified under parts/bundle/prime configs.
+
+    This is a intermediate structure until we have the full Lifecycle in place.
+    """
+
+    @classmethod
+    def from_dict(cls, parts):
+        """Build from a dicts sequence."""
+        prime = parts.get('bundle', {}).get('prime', [])
+
+        # validate that all are relative
+        for spec in prime:
+            # check if it's an absolute path using POSIX's '/' (not os.path.sep, as the charm's
+            # config is independent of the platform where charmcraft is running)
+            if spec[0] == '/':
+                raise CommandError(
+                    "Bad charmcraft.yaml content; the paths specifications in "
+                    "'parts.bundle.prime' must be relative: found {!r}.".format(spec))
+        return cls(prime)
+
+
+@attr.s(kw_only=True, frozen=True)
+class _Project:
+    """Configuration for all project-related options, used internally."""
+
+    dirpath = attr.ib(default=None)
+
+
+@attr.s(kw_only=True, frozen=True)
+class _Config:
```

Comment:
out of curiosity why lead with '_'?

Snapcraft is all over the place with its naming convention, but generally I'd expect if a class or function is meant to be used external to the module, it should not be labeled as such.  And I'd expect there to be other usage of Config, Project, etc. but maybe not :D

Reply:
My idea was to indicate the possible developers that they are not expected to instantiate that class directly, but using the module's `load` function (which receives the directory where the config is, and stores that in the config itself, a value that is critical for some commands).

Reply:
If anyone outside the module is going to use an interface, it should not be marked as internal.

Since load() is returning a Config object, presumably for external use, the Config object itself should be public.

FWIW load() may be encapsulated as a classmethod in Config if it's strictly to init the Config.

Also, consider the case where you may want to create a Config() object without a backing yaml (helpful in tests).

Reply:
It was originally a classmethod, I was requested to decouple it from the class in the draft branch. In any case, it's not a problem for the tests.

I'll make the classes public, though.